### PR TITLE
gix-merge: support SHA-256 and fix a file/directory merge bug

### PR DESCRIPTION
**Pull request by 10ne1**
**What**

This makes gix-merge work under SHA-256, running its tree-merge baselines under sha256.
This surfaced a pre-existing bug in the merge itself, which this also fixes (see details in the first commit message).

I can't say I'm entirely happy with the `tests/fixtures/tree-baseline.sh` in the 3rd commit, which I consider overly verbose for the harcoded hashes, however it's the best myself and Claude could come up with.

This is part of #281 .

**Testing**
gix-merge now runs under both hashes; added to the justfile:
```
  GIX_TEST_FIXTURE_HASH=sha1 cargo nextest run -p gix-merge --no-fail-fast
  GIX_TEST_FIXTURE_HASH=sha256 cargo nextest run -p gix-merge --no-fail-fast
```

**AI disclosure**
This PR was written with the help of Claude Code, but was **not** vibe-coded. I'm not a fan of vibe-coding. I used AI to better understand the codebase myself (asked it a lot of questions), review both my code and the code generated by AI, used it especially to detect bugs and corner-cases and suggest fixes, however I very carefully reviewed & edited each code/comment lines, the commit messages and so on, until I was satisfied with the result.

### CODE REVIEW

File: gix-merge/tests/merge/tree/baseline.rs [hunk 1]
```diff
@@ -168,8 +168,17 @@ impl Iterator for Expectations<'_> {
...
+            gix_odb::store::init::Options {
+                object_hash,
+                ..Default::default()
+            },
+        )
+        .expect("object dir exists");
+        let objects = gix_odb::memory::Proxy::new(odb, object_hash);
         let our_commit_id = gix_hash::ObjectId::from_hex(our_commit_id.as_bytes()).unwrap();
         let their_commit_id = gix_hash::ObjectId::from_hex(their_commit_id.as_bytes()).unwrap();
         let merge_info = parse_merge_info(std::fs::read_to_string(subdir_path.join(merge_info_filename)).unwrap());
```
[the remark this answers is not in the document]
Comment (10ne1) [reply gix-merge/tests/merge/tree/baseline.rs:184]: These were not introduced by my commit... I try to not modify unrelated code and keep diffs minimal.

Perhaps it's better to do a separate "unwrap cleanup" PR.

File: gix-merge/tests/merge/tree/baseline.rs [hunk 2]
```diff
@@ -168,8 +168,17 @@ impl Iterator for Expectations<'_> {
             "diff3" => ConflictStyle::Diff3,
             unknown => unreachable!("Unknown conflict style: '{unknown}'"),
         };
-        let odb = gix_odb::at(subdir_path.join(".git/objects")).expect("object dir exists");
-        let objects = gix_odb::memory::Proxy::new(odb, gix_hash::Kind::Sha1);
+        let object_hash = gix_testtools::object_hash();
+        let odb = gix_odb::at_opts(
+            subdir_path.join(".git/objects"),
+            Vec::new(),
```
[the remark this answers is not in the document]
Comment (10ne1) [reply gix-merge/tests/merge/tree/baseline.rs:174]: Fair point

**Review by Byron — APPROVED** [review]
Thanks a lot for your help *and* the fix, that is a great find.

Let me share an important note to make this less puzzling: I have removed the updated `tar` files from your commits and am re-adding these in the review commit. The background is that these files are harder to review than their source, and as derived value it should be produced by the reviewer to ensure source and derivative match.

I also added a note that states `gix-merge` shouldn't be dependent on assumed merge order in rename tracking right now, while there is hope one day it could if that's faster, and once the implementation caught up with Git.

File: gix-merge/src/tree/function.rs [hunk 3]
```diff
@@ -700,21 +700,22 @@ where
...
-                                            change.inner.entry_mode().is_tree()
-                                                && change.inner.location() == location
-                                                && matches!(change.inner, Change::Addition { .. })
-                                        })
-                                        .unwrap_or_default()
+                                    changes.iter().any(|change| {
+                                        change.inner.entry_mode().is_tree()
+                                            && change.inner.location() == location
+                                            && matches!(change.inner, Change::Addition { .. })
+                                    })
```
[the remark this answers is not in the document]
Comment (Byron) [reply gix-merge/src/tree/function.rs:715]: I think it's Ok, and for good measure I reordered the comparisons from fast to slow.

File: gix-merge/tests/merge/tree/baseline.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Byron) [reply gix-merge/tests/merge/tree/baseline.rs:174]: Typical Copilot - let's keep it simple.

File: gix-merge/tests/merge/tree/baseline.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Byron) [reply gix-merge/tests/merge/tree/baseline.rs:184]: Copilot is looking for work in places that didn't change isn't great :/.

**Comment by 10ne1** [comment]
> The background is that these files are harder to review than their source, and as derived value it should be produced by the reviewer to ensure source and derivative match.

Thank you @Byron I 100% agree and will stop adding those generated files in future PRs, I will just leave a reminder note in the PR description for the reviewer to generate & add the files.

**Comment by Byron** [comment]
Actually, please feel free to add them as CI won't pass without them (or if they are outdated). I will take care of it then.
